### PR TITLE
2180 mongo event repository

### PR DESCRIPTION
**Pull request by cakekoa**
# What does this PR do?

Fixes part of  #2180.

Add a repository for storing events in MongoDB.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] ~~Was the CHANGELOG.md updated to reflect the changes?~~
* [ ] ~~Was the documentation framework updated to reflect the changes?~~
* [ ] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Just ran the unit tests
* [ ] ~~If applicable, add screenshots or log transcripts of the feature working~~

### CODE REVIEW

File: monkey/monkey_island/cc/repository/mongo_event_repository.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        self._events_collection = mongo_client.monkey_island.events
+        self._serializers = serializer_registry
+
+    def save_event(self, event: AbstractAgentEvent):
+        try:
+            serializer = self._serializers[type(event)]
+            serialized_event = serializer.serialize(event)
+            self._events_collection.insert_one(serialized_event)
+        except Exception as err:
+            raise StorageError(err)
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/repository/mongo_event_repository.py:27]: Missing a message

File: monkey/monkey_island/cc/repository/mongo_event_repository.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    def get_events(self) -> Sequence[AbstractAgentEvent]:
+        try:
+            serialized_events = list(self._events_collection.find())
+            return list(map(self._deserialize, serialized_events))
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/repository/mongo_event_repository.py:32]: I think `serialized_events = list(...` is unnecessary. `map()` takes an `Iterable` and `find()` returns a `Cursor` which I believe is iterable.

File: monkey/monkey_island/cc/repository/mongo_event_repository.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            raise RetrievalError(f"Error retrieving events for source {source}: {err}")
+
+    def reset(self):
+        try:
+            self._events_collection.drop()
+        except Exception as err:
+            raise RemovalError(f"Error resetting the repository: {err}")
+
+    def _deserialize(self, mongo_record: MutableMapping[str, Any]) -> AbstractAgentEvent:
+        del mongo_record[MONGO_OBJECT_ID_KEY]
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/repository/mongo_event_repository.py:68]: You don't need to `del` the ID if you add `{'_id': False}` to your `find()` call. See [this](https://github.com/guardicore/monkey/blob/0b5dd3832b7591cef77a3a24cd3802c1be04fb55/monkey/monkey_island/cc/repository/mongo_machine_repository.py#L72-L79) for an example.

File: monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    return MongoEventRepository(mongo_client, event_serializer_registry)
+
+
+@pytest.fixture
+def error_raising_mongo_client(mongo_client) -> mongomock.MongoClient:
+    mongo_client = MagicMock(spec=mongomock.MongoClient)
+    mongo_client.monkey_island = MagicMock(spec=mongomock.Database)
+    mongo_client.monkey_island.events = MagicMock(spec=mongomock.Collection)
+
+    # The first call to find() must succeed
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py:64]: Is this true or just copy/paste?

File: monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    mongo_client.monkey_island.events.find = MagicMock(
+        # side_effect=chain([MagicMock()], repeat(Exception("some exception")))
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py:66]:
```suggestion
```

File: monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        side_effect=Exception("some exception")
+    )
+    mongo_client.monkey_island.events.find_one = MagicMock(side_effect=Exception("some exception"))
+    mongo_client.monkey_island.events.insert_one = MagicMock(
+        side_effect=Exception("some exception")
+    )
+    mongo_client.monkey_island.events.replace_one = MagicMock(
+        side_effect=Exception("some exception")
+    )
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py:75]:
```suggestion
```

File: monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    mongo_client = MagicMock(spec=mongomock.MongoClient)
+    mongo_client.monkey_island = MagicMock(spec=mongomock.Database)
+    mongo_client.monkey_island.events = MagicMock(spec=mongomock.Collection)
+
+    # The first call to find() must succeed
+    mongo_client.monkey_island.events.find = MagicMock(
+        # side_effect=chain([MagicMock()], repeat(Exception("some exception")))
+        side_effect=Exception("some exception")
+    )
+    mongo_client.monkey_island.events.find_one = MagicMock(side_effect=Exception("some exception"))
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py:69]:
```suggestion
```

File: monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    with pytest.raises(RetrievalError):
+        source_event = EVENTS[2]
+        error_raising_mongo_repository.get_events_by_source(source_event.source)
+
+
+def test_mongo_event_repository__reset(mongo_repository: IEventRepository):
+    mongo_repository.reset()
+    events = mongo_repository.get_events()
+
+    assert not events
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py:172]: I like to test that the repository has something, then reset, then test that it's empty. Maybe it's overkill, but it prevents the test from passing erroniously.

File: monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py (hunk 4 above)
Comment (cakekoa) [reply monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py]: Not true, just a residual comment from borrowed code

File: monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py (hunk 8 above)
Comment (cakekoa) [reply monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py:164]: The fixture already has all of the events from `EVENTS`. Or would you like events to be explicitly added in the test?

File: monkey/monkey_island/cc/repository/mongo_event_repository.py (hunk 2 above)
Comment (cakekoa) [reply monkey/monkey_island/cc/repository/mongo_event_repository.py]: It returns an [`Iterator`](https://docs.python.org/3/library/functions.html#map)

Comment (mssalvatore) [reply monkey/monkey_island/cc/repository/mongo_event_repository.py]: `map()` does. I'm saying `find()` does also, so you don't need to `list(self._events_collection.find())`.

File: monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py (hunk 8 above)
Comment (mssalvatore) [reply monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py:164]: I prefer to explicitly `get_events()`, make sure it's non-zero, and then call `reset()`.

Comment (cakekoa) [reply monkey/tests/unit_tests/monkey_island/cc/repository/test_mongo_event_repository.py:167]: Thanks for the clarification!

File: monkey/monkey_island/cc/repository/mongo_event_repository.py (hunk 2 above)
Comment (cakekoa) [reply monkey/monkey_island/cc/repository/mongo_event_repository.py]: Aha! Gotcha

**Pull request merged**